### DISCUSSION TITLE
Destination Clickhouse - Add Typing and Deduping

### DISCUSSION THREAD

**Discussion by jbfbell** [airbytehq/airbyte#35339] [Connector Ideas and Features]
We [removed](https://github.com/airbytehq/airbyte/issues/32528) ([PR](https://github.com/airbytehq/airbyte/pull/34637) )"normalization" from Destination Clickhouse. Normalization was a process which would extract data from the raw json format and cast the data into tables with the appropriate types. For other destinations, we implemented [Typing and Deduping](https://airbyte.com/blog/introducing-airbyte-destinations-v2-typing-deduping) to replace normalization and we will bring typing and deduplication to Clickhouse if there is sufficient interest

**The purpose of this discussion is to gauge interest in bringing this functionality back to Clickhouse.** 

Without typing and deduping, you would need to integrate another tool, such as [dbt](https://clickhouse.com/docs/en/integrations/dbt) to accomplish the same result. While integrating another tool gives you greater customization over the data format, it comes with additional operational overhead.

Should Airbyte bring Typing and Deduping to Clickhouse?

# Update

Please check the new connector Clickhouse and the beta test discussion happening [here](https://github.com/airbytehq/airbyte/discussions/62035)

**Comment by asimumba** [comment]
How can I revert to pre 1.0.0 upgrade of the Clickhouse destination?

**Comment by kvosha** [comment]
Definitely we need normalization for clickhouse! Is there any way to stay on old version of clickhouse connection until normalization will be implemented?

**Reply by alberk8** [reply]
Yes, bring Typing and Deduping ClickHouse

**Reply by Riya-gupta-8** [reply]
Yes . You can degrade to v0.2.5 version for clickhouse

**Comment by nailele** [comment]
airbyte should bring typing and deduping for clickhouse 🥹

**Reply by nailele** [reply]
![image](https://github.com/user-attachments/assets/1838589b-dd48-47f5-a0c0-8d1a702364d2)

**Comment by MaxMusing** [comment]
Strong +1 on this, we would benefit greatly from typing and deduping on ClickHouse as a destination.

**Comment by asimumba** [comment]
In support to have this feature implemented.

**Comment by machariamuguku** [comment]
FYI: The airbyte team published the connector docker images on dockerhub. You can 'revert' to an older connector version (with normalization for example) by uploading it's older image as a custom connector. See clickhouse connector images [here](https://hub.docker.com/r/airbyte/destination-clickhouse/tags) cc @asimumba @kvosha

**Reply by asimumba** [reply]
Hi, I actually resolved this already. Thanks.

**Reply by Aryanshaw** [reply]
how did u solve this problem can u tell me

**Comment by AlexisSerneels** [comment]
I've spent hours debugging why the final tables were not created before finding this discussion.
We definitely need this feature ;)

**Comment by vietpm** [comment]
I can't understand why this feature is removed. Please bring this back in destination v2.

For anyone wants this feature, you can go back to version 0.2.5 of clickhouse destination. Note that you'll be switched back to destination v1.
1, Settings -> Destinations -> Clickhouse. Change version to 0.2.5 and click "Change".
2, Enter your connection, choose Transformation, change to "Normalized tabular data", click "Save". 
3, (optional) If you're using Incremental sync mode, "Incremental | Append + Deduped" is available now so you may be want to change. 
4, Reset data, resync from scratch.

**Reply by shakalandy** [reply]
@vietpm does this still work for you?
The "Transformation" tab is not shown anymore, not sure if it's because of a new ClickHouse Version?
Downgrading to ClickHouse Connector 0.2.4 and then back to 0.2.5 does not bring transformation back anymore...

**Reply by shakalandy** [reply]
ok, Downgraded Airbyte to 0.63.3, works again - with 0.63.9 the "transformation" tab was lost despite the downgrade of the clickhouse Destination.

**Reply by Aryanshaw** [reply]
Hey @shakalandy does it still work for you?

**Reply by shakalandy** [reply]
Migrated to MySQL as Destination in this case.

**Comment by edwardysun** [comment]
+1 need this

**Comment by cedstrom** [comment]
Holy hell, please bring normalization back +1000.

**Comment by luishdosreis** [comment]
Pretty please bring normalisation back :) 🍰

**Comment by zowen-ch** [comment]
We need the normalization for our workflows. Please keep supporting this in Clickhouse!

**Comment by georgeemr** [comment]
Please bring it back!

**Comment by jesperbagge** [comment]
A big heartfelt YES! Please bring Typing and Deduping to ClickHouse!

**Comment by Lawchak-Fadhil** [comment]
Please bring it back, I've spent countless hours trying to figure out what is wrong with my setup! We really need this for ClickHouse.

**Comment by alekseyp** [comment]
Just for the history - as of now - this is the number one issue by popularity in https://github.com/airbytehq/airbyte/discussions/categories/ideas-and-features

**Comment by lfigueira** [comment]
+1 - this would be extremely handy

**Comment by tamilselvan-stratforge** [comment]
+1 - Definitely will need this feature for Clickhouse to reduce the operational overhead.

**Reply by dpurtov** [reply]
Even more - this [commit](https://github.com/airbytehq/airbyte/pull/27958) was extremely useful. Please bring normalization back and merge commit above.

**Comment by prashant-ux23** [comment]
Please bring normalization back for Clickhouse +1

**Comment by matheuzinoficial** [comment]
+1 - We are waiting for this for too long, it would be very good for our company!

**Comment by o1lab** [comment]
Please bring normalization back for Clickhouse +1

**Comment by ThomasSertorio** [comment]
+1 - Please, please, we want it 😎
Really like to see normalization back for CH.

**Comment by prakashbalaji** [comment]
+1 We would need this feature back.

**Comment by CNimmo16** [comment]
+1 like many others I've spent hours trying to figure out what's going on, please bring this back!

**Comment by vitas-pm** [comment]
I just don't get it. Without trying to sound rude but Airbyte as a whole is completely useless for our purposes without normalization. Why take the T out of the ETL only to ask if you should put it back in? This downgrade has completely messed up our pipelines and cost us multiple days to fix. Please, please put it back in asap.

**Comment by evantahler** [comment]
An update from the Airbyte engineering team:
* Airbyte made some rather big changes to our destinations in the past year. As Clickhouse was not a certified connector, not was it as popular as Snowflake, Bigquery, etc, it was not included in the original batch of connectors that was migrated to [Destinations V2](https://airbyte.com/blog/introducing-airbyte-destinations-v2-typing-deduping), and now [Refreshes](https://docs.airbyte.com/operator-guides/refreshes).  As we have changed how the Airbyte Platform works (e.g. removing the old dbt-based normalization), we had to remove the old normalization system from the Clickhouse destination so it would still function.  
  * We removed dbt-based normalization to dramatically increase reliability ([e.g.](https://airbyte.com/blog/record-change-history)) and performance.
* At the moment, we are focused on revamping our tools so that we can build and certify connectors _faster_ going forward.  This includes building a new Bulk Load (Destinations) CDK which will include adding Typing and Deduping, Refrehses, etc to a destination easily.  Today, it's a tricky thing to do.
* The plan is to complete our new Bulk Load (Destinations) CDK in late 2024, and re-introduce the community to this new tool in early 2025.  At that point, we will welcome community contributions to modernize the connector, which will include adding Typing and Deduping and Refreshes support.  Or, at that time, if we've gotten significant interest from our commercial users (Airbyte Cloud or Self-hosted Airbyte Enterprise), we will likely in-house the connector and certify it.

**Comment by SergeAx** [comment]
Hello, and thank you for your work on Airbyte, especially for making it open-source and available for enthusiasts and low-budget builders.

What is the current status of the Clickhouse destination connector? From the [documentation page](https://docs.airbyte.com/integrations/destinations/clickhouse) it seems that all the features are currently supported: four "yes" features and zero "No". However, I cannot make it support normalization and transformation of data.

Please note that Clickhouse is also an open-source product and thus is cherished by enthusiasts and makers, the crowd you are addressing by making Airbyte OSS.

**Comment by ankitswifter** [comment]
Yes Please, bring typing and deduping for clickhouse.

**Comment by hiteshchinu** [comment]
Yes we need it back!!!!!!!!!!!!

**Comment by sky93** [comment]
same problem here airbyte isn't useful for clickhouse anymore.

**Comment by gustavfjorder** [comment]
Can we fix the docs? They currently say that dedupe mode is supported https://docs.airbyte.com/integrations/destinations/clickhouse

**Comment by Aryanshaw** [comment]
guys do not rely on airbyte I did it I figured out a way to normalize it but it is using dbt so instead of ETL you guys will have to rely on ELT
I wrote an article on this tutorial u can try it
https://medium.com/@aryanrot234/airbyte-raw-data-transformation-using-dbt-dynamically-221de752733b

**Comment by geekbleek** [comment]
+1 this is drastically needed. ETL without the T just becomes the.  Clickhouse (self-hosted and cloud) is one of the fastest performing columnar stores for analytics.  If we're evaluating entire new ETL/schedulers/implementing ELT we might as well eval moving off Airbyte (self hosted or enterprise).

And ELT isn't the solve, because without dedupe, streams with high velocity of change will result in large raw datasets with many duplicates.  Yes you can ELT to new tables and sort/dedupe on primary keys, but that reinforces the need for standing up a job scheduler stack with dbt where it wasn't required at all previously.

**Reply by Aryanshaw** [reply]
checkout my article

**Comment by mert-ozturk-lgs** [comment]
Yes, please integrate Typing and Deduping to Clickhouse. It is essential. Without the functionality, Airbyte becomes less attractive.

**Reply by Aryanshaw** [reply]
checkout my article

**Reply by mert-ozturk-lgs** [reply]
Yes, I saw your article, it's good! But still it would be better to avoid the additional effort in future.

**Comment by blarghmatey** [comment]
Is there any information available about whether and when the Clickhouse destination will be updated to support the typing and deduping functionality? I think this is a rather critical feature across Airbyte and should be the default behavior for all supported destinations.

**Reply by philippevezina** [reply]
I just saw that it's on their roadmap for Feb. - May 2025 so I am guessing it's reasonable to exepect more information around that time.

**Comment by zackrspv** [comment]
any other good alternatives until this is actually working again -.-

**Reply by Aryanshaw** [reply]
I made a solution on how to transform the raw airbyte file using DBT and it is 100% dynamic just provide the db name and run the script and everything will be done

ELT pipeline between source to airbyte to clickhouse
https://medium.com/@aryanrot234/airbyte-raw-data-transformation-using-dbt-dynamically-221de752733b

**Comment by geekbleek** [comment]
Any update here from the Airbyte team?  Great that there's lots of other scripted solutions, but if we're setting up dbt etc and running them in schedulers like Airflow, might as well setup us dlt etc, and no longer use Airbyte if Clickhouse is a destination.

**Reply by jesperbagge** [reply]
That’s exactly what I did, in the end. Running dlt and dbt on Airflow gave us less tools to work with and roughly a 500% increase in performance due to how fast dlt runs with pyarrow as backend - and in parallel! - compared to old Java-connectors that does one table at a time.

**Comment by Segfaultd** [comment]
Any chance this get attention? No way for us to keep using Airbyte if its not handled.

**Comment by alexworkingcats** [comment]
Yes please this would be great !

**Comment by john-griffin** [comment]
Please add support for ClickHouse.

**Comment by Aryanshaw** [comment]
Hey guys, did you find any other alternatives for airbyte? (free open source)

**Reply by Segfaultd** [reply]
Not using AirByte at all. They don't seem to care about something broken for months that was perfectly working before.
Made the move to Meltano self-deployed.

**Reply by jesperbagge** [reply]
We migrated to [dlt](https://dlthub.com). Orders of magnitude faster _and_ can move multiple tables in parallel.

**Reply by john-griffin** [reply]
Also moving to Meltano

**Reply by Aryanshaw** [reply]
tell me is meltano faster? and does it provide rest API access like airbyte? or is it just cli based?

**Reply by Segfaultd** [reply]
https://docs.meltano.com/

**Comment by yalattas** [comment]
+1

**Comment by mikich-trintech** [comment]
# +1
### I'm `surprised` and `disappointed` to see Airbyte abandoning its efforts in supporting ClickHouse!
ClickHouse engine is gaining significant traction within the community. 
Meanwhile, ClickHouse is experiencing rapid growth, with an increasing number of users and contributions.

**Comment by ian-at-airbyte** [comment]
Hi, everyone! We're working on ClickHouse this cycle. This is one of our most upvoted requests ever and we definitely haven't forgotten about it! I won't speculate on a delivery date at this early stage, but I do hope to be able to do said speculation soon.

👨‍💻 👩‍💻 🧑‍💻

**Reply by abdulfaqihalm** [reply]
Thanks @ian-at-airbyte! Any update from the Airbyte team regarding the delivery?

**Reply by ian-at-airbyte** [reply]
No specific delivery date yet. For sure not this month, though. Will keep you posted.

**Reply by alepietrobon** [reply]
adding another solid +1. Looking forward to the upgraded connector. Thanks!

**Reply by sarthakgarg91** [reply]
++1 urgently need this

**Reply by havrilyukv** [reply]
also urgently need this :(

**Reply by mekedron** [reply]
Thanks @ian-at-airbyte!

I see it's being actively developed right now:
https://github.com/airbytehq/airbyte/branches/all?query=clickhouse

https://hub.docker.com/r/airbyte/destination-clickhouse-v2

Thanks you very much!

**Reply by havrilyukv** [reply]
> No specific delivery date yet. For sure not this month, though. Will keep you posted.

Hey! 

Can we expect that CHv2 will be ready in June?

**Reply by ian-at-airbyte** [reply]
It seems like we're currently on track for a delivery window sometime in June or July. I'll try to make that more specific shortly.

**Reply by havrilyukv** [reply]
> It seems like we're currently on track for a delivery window sometime in June or July. I'll try to make that more specific shortly.

thanks for update. really want to stick with your software but it is must have for us.

**Reply by ian-at-airbyte** [reply]
Thanks for your patience. This one has been a long time coming but it was important to us to get it done.

**Reply by lerminou** [reply]
First official version available: https://docs.airbyte.com/integrations/destinations/clickhouse-v2 
I will go try it, thank you for your work

**Reply by havrilyukv** [reply]
> First official version available: https://docs.airbyte.com/integrations/destinations/clickhouse-v2 I will go try it, thank you for your work

do not see it in my Airbyte Open Source marketplace (((((

**Comment by lerminou** [comment]
Hi, do you plan to support connecting to a ClickHouse cluster in your roadmap? @ian-at-airbyte 
an old issue: https://github.com/airbytehq/airbyte/issues/9134

**Reply by ian-at-airbyte** [reply]
Hi @lerminou. This is not planned as part of the current typing and deduping project, but I've flagged it with the dev team so they're aware.

**Comment by marcosmarxm** [comment]
Please check the new connector Clickhouse and the beta test discussion happening [here](https://github.com/airbytehq/airbyte/discussions/62035)
